### ISSUE TITLE
Error CS8103 Combined length of user strings used by the program exceeds allowed limit. Try to decrease use of string literals.

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#85413]
I receive this error on my datasets since I've installed the latest VS2017. I have big tables, and think the MSDataSetGenerator generates code different, so now the compiler can't handle the big generated strings anymore?

There is not much information about the error message BC37255.

Can you help?

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/192728/bc37255-combined-length-of-user-strings-used-by-th.html
VSTS ticketId: 561492_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

**Comment by joshyates1980** [comment]
`    <Features>$(Features);experimental-data-section-string-literals</Features>`

well, this is amazing. Does this mean I do not have to create resource strings for my larger string literals?
Much gratitude to @jjonescz and @DamianEdwards 
- josh

**Comment by DamianEdwards** [comment]
@joshyates1980 glad it's working for you! Yes, in theory this means you can simply declare and use string literals without worrying about their size.

**Comment by dafemec** [comment]
I am encountering the same error in Visual Studio 2022, 17.14.15 (September 2025), but I am attempting to build a Visual-Basic application.  I have tried inserting the specified line in the appropriate .vbproj file inside PropertyGroup but to no avail.  Any suggestions?  Thanks in advance for any help offered!

<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <RootNamespace>VisualBasicFileHandling</RootNamespace>
    <TargetFramework>net8.0</TargetFramework>
    <Features>$(Features);experimental-data-section-string-literals</Features>
  </PropertyGroup>

</Project>

**Comment by jjonescz** [comment]
The feature is currently only implemented for C#, not VB.

**Comment by peng5545** [comment]
@jjonescz 

I have a large MVC project that I'm currently upgrading from .NET Framework 4.8 to .NET 8. It contains thousands of .cshtml files, many of which contain JavaScript. I can't split the project as the primary target for the upgrade, and I can't migrate the JavaScript code out. In .NET Framework 4.8, I used RazorGenerator.MsBuild to precompile the .cshtml, which worked fine with experimental-data-section-string-literals. However, in .NET 8, whenever I use reflection, such as 
`Assembly.GetExecutingAssembly().GetTypes()`
 I get the following error: 
`"Internal limitation: too many fields.": "<PrivateImplementationDetails>". `
Is there any solution? Thank you.

.NET SDK:
 Version:           9.0.305
 Commit:            3fc74f3529
 Workload version:  9.0.300-manifests.6fcb754b
 MSBuild version:   17.14.21+8929ca9e3

Visual Studio  17.14.16

**Comment by CyrusNajmabadi** [comment]
@peng5545 please open a new issue.  Having a copy of your project would also be very helpful.  If that can't be linked here, you can open an issue on developercommunity and link it private there.

**Comment by jjonescz** [comment]
I can repro that - filed https://github.com/dotnet/roslyn/issues/80494 to track that issue. Thanks.

The workaround might be tweaking the feature flag, e.g. setting it to 100, 200, etc. (doesn't have to be a round number) like

```csproj
<Features>$(Features);experimental-data-section-string-literals=200</Features>
```

until you don't get any errors.

**Comment by peng5545** [comment]
<img width="1678" height="408" alt="Image" src="https://github.com/user-attachments/assets/eea1a84b-7c0f-4016-9231-b2171eff5f33" />

Channel:  Stable
Version: 18.0.0

<Features>$(Features);experimental-data-section-string-literals=110</Features>
Is any change with vs 2026,not working now.if change number to less or equal to 100 can work.
vs2022 17.14.20 work well.

@jjonescz

**Comment by HomeGrownCoder** [comment]
After updating to Visual Studio 2022 17.14.20, we are now running into the issue with our VB.NET solution. Seems that the work around is only for C#, so what is the recommended path for VB projects, aside from refactoring strings from our legacy code into resource files?

**Comment by CyrusNajmabadi** [comment]
@HomeGrownCoder you could move your strings into a C# project.

**Comment by HomeGrownCoder** [comment]
> [@HomeGrownCoder](https://github.com/HomeGrownCoder) you could move your strings into a C# project.

That's an option if we have to...just extra work and testing.

As a follow-up, we just tested on another machine in our office with VS 2022 17.14.20 and it does not have the issue. One difference between the machines is the one with the issue also has VS 2026 installed, while the unaffected machine does not have VS 2026. Is there any chance that installation of 2026 updates components (i.e. compiler) that VS 2022 is using?

**Comment by jjonescz** [comment]
> Is there any chance that installation of 2026 updates components (i.e. compiler) that VS 2022 is using?

Installation of VS can bring in new versions of .NET SDK and that's what is actually used during building (usually the latest globally  installed .NET SDK is used) - you can check `dotnet --list-sdks`. To pin a specific SDK version, you can use `global.json`.

**Comment by blackbaud-rossthomson** [comment]
@jjonescz - was the max string limit changed between 2022 and 2026? I've got a .Net Framework 4.8 vbproj that I can quite happily build on 2022, but that throws a near endless succession of “error BC37255: Combined length of user strings used by the program exceeds allowed limit. Try to decrease use of string or XML literals.” When I build in 2026.

I’m building in 2022 and 2026 on the same machine, 2022 builds 2026 does not?

**Comment by jjonescz** [comment]
No, the limit was not changed. It's quite possible you were near the limit before and some changes in the compiler caused it to emit more strings or strings in a different order. If you can provide a repro (perhaps using [complog](https://github.com/jaredpar/complog?tab=readme-ov-file#creating-compiler-logs)), I could try taking a deeper look at what's happening (you can share a repro privately by logging a feedback ticket from visual studio and linking it back here).

**Comment by jjonescz** [comment]
Btw, I created the following issue to track support of the `experimental-data-section-string-literals` feature for VB, please upvote it if you want it, so we can track the demand, thanks.

- https://github.com/dotnet/roslyn/issues/81313

**Comment by blackbaud-rossthomson** [comment]
> No, the limit was not changed. It's quite possible you were near the limit before and some changes in the compiler caused it to emit more strings or strings in a different order. If you can provide a repro (perhaps using [complog](https://github.com/jaredpar/complog?tab=readme-ov-file#creating-compiler-logs)), I could try taking a deeper look at what's happening (you can share a repro privately by logging a feedback ticket from visual studio and linking it back here).

Sure thing - added here:

https://developercommunity.visualstudio.com/t/Visual-Studio-2026:-Exception-error-BC3/11002044

**Comment by CePur** [comment]
Just installed 2026 Version: 18.1.0 and getting this error. Nothing changed in solution. It was building happily in 2022 just before installing 2026. Now, both throws this exception.

https://developercommunity.visualstudio.com/t/Visual-Studio-2026:-Exception-error-BC3/11015804

**Comment by jjonescz** [comment]
@CePur I've responded on that ticket requesting more information.

**Comment by robinwilson16** [comment]
> Roll on .NET 9! Please could this be fixed as it has been 6 years now since it worked fine and was not an issue, then it was broken and has remained broken for the 6 years and the only fix is to have smaller .NET projects or put content into databases or resource files or split the project into multiple projects. Depending on the nature of the application this is not always feasible or adds unneeded complexity.
> 
> It's like someone at Microsoft breaking Microsoft Word so you could no longer write more than 1,000 characters in a single document and the workaround being to create a separate document for each chapter or embed documents into each other. I doubt that would remain unpatched for 6 years.
> 
> Please fix this so the size of projects does not need to be restrained.

Please can this be fixed as it's been over another 1.5 years and is clearly affecting lots of people from the comments and 2 NET versions later it has become more of a problem if anything, so this is over 7.5 years waiting for a fix and having to break up larger projects into multiple smaller projects to get around the hard limitation.

**Comment by jjonescz** [comment]
> Please can this be fixed as it's been over another 1.5 years and is clearly affecting lots of people from the comments

@robinwilson16 We have a fix for this - https://github.com/dotnet/roslyn/blob/main/docs/features/string-literals-data-section.md - is that not working for you?

**Comment by robinwilson16** [comment]
Thanks I wasn't aware a fix had been released with it being under the feature flag.
I have hit this on lots of occasions but am not currently building a sufficiently large project that triggers this issue but when I do I will ensure I turn on the feature flag.
Will it eventually default to turned on once more fully tested or will it be like long paths support in Windows where you always have to remember to turn it on.
It's great there is now a solution.

Thanks
Robin

**Comment by jjonescz** [comment]
> Will it eventually default to turned on once more fully tested or will it be like long paths support in Windows where you always have to remember to turn it on.

Yes, having this enabled by default eventually is the plan.
